### PR TITLE
new implementation for git pull

### PR DESCRIPTION
**Pull request by lazytesting**
- Backup of files before clearing the /config folder
- Copy back all non YAML files and the secrets.yaml after the git clone
- More verbose error handling. Also logging the GIT exceptions.
- Splitted code out into functions
- Check SSH connection before setting the key

### CODE REVIEW

**Comment by homeassistant** [comment]
Hi @lazytesting, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;hassio-addons#218).

Once you do that we will be able to review and accept this pull request.

Thanks!

**Comment by lazytesting** [comment]
build is failing on apt-get, I did a quick investigation and it seems to be related to this issue: https://github.com/travis-ci/travis-ci/issues/4838

File: git_pull/run.sh [hunk 1]
```diff
@@ -9,53 +10,83 @@ REPOSITORY=$(jq --raw-output '.repository' $CONFIG_PATH)
...
+function git-clone {
+    # back to main folder
+    cd ..
 
-# init config repositorie
-if [ ! -d /config/.git ]; then
-    echo "[Info] cleanup config folder and clone from repositorie"
-    rm -rf /config/.[!.]* /config/* 2&> /dev/null
+    # create backup
+    BACKUP_LOCATION="./tmp/config-$(date +%Y-%m-%d_%H-%M-%S)"
```
Comment (pvizeli) [inline git_pull/run.sh]: `/tmp/...`

File: git_pull/run.sh [hunk 2]
```diff
@@ -9,53 +10,83 @@ REPOSITORY=$(jq --raw-output '.repository' $CONFIG_PATH)
...
+    echo "Backup configuration to $BACKUP_LOCATION"
+    
+    mkdir "${BACKUP_LOCATION}" || { echo "[Error] Creation of backup directory failed"; exit 1; }
+    cp -rf config/* "${BACKUP_LOCATION}" || { echo "[Error] Copy files to backup directory failed"; exit 1; }
```
Comment (pvizeli) [inline git_pull/run.sh]: `/config/*`

File: git_pull/run.sh [hunk 3]
```diff
@@ -9,53 +10,83 @@ REPOSITORY=$(jq --raw-output '.repository' $CONFIG_PATH)
...
+    
+    # remove config folder content
+    rm -rf config/{,.[!.],..?}* || { echo "[Error] Clearing /config failed"; exit 1; }
```
Comment (pvizeli) [inline git_pull/run.sh]: Absolut path

File: git_pull/run.sh [hunk 4]
```diff
@@ -9,53 +10,83 @@ REPOSITORY=$(jq --raw-output '.repository' $CONFIG_PATH)
...
-if [ ! -z "$DEPLOYMENT_KEY" ]; then
-
+#### functions ####
+function add-ssh-key {
+    echo "Start adding SSH key"
     mkdir -p ~/.ssh
-    echo "[Info] disable StrictHostKeyChecking for ssh"
+
     echo "Host *" > ~/.ssh/config
     echo "    StrictHostKeyChecking no" >> ~/.ssh/config
```
Comment (pvizeli) [inline git_pull/run.sh]:
```
(
  echo "xy"
  echo "z"
) > ~/.ssh/config
```

**Comment by pvizeli** [comment]
Good work 👍

**Comment by lazytesting** [comment]
thanks @pvizeli ! Will implement your PR comments and test it ASAP. Do you have an idea how to fix the CI tests?

**Comment by pvizeli** [comment]
I need rebase you branch with last master

**Comment by pvizeli** [comment]
Please address the lint issue

**Comment by pvizeli** [comment]
Good work :+1:

**Pull request merged**